File: 3.x.x/crates-utopia/messaging/src/adapter/email/sendgrid.rs
//! PHP `Utopia\Messaging\Adapter\Email\Sendgrid`.

use serde_json::{json, Value};

use super::{MAX_ATTACHMENT_BYTES, TYPE};
use crate::adapter::{expect_email, Adapter, AdapterBase, SendResult};
use crate::error::MessagingError;
use crate::message::{Message, MessageKind};
use crate::messages::Email;
use crate::php::php_empty;
use crate::response::{Response, ResponseData};

/// PHP `Adapter\Email\Sendgrid`.
#[derive(Debug)]
pub struct Sendgrid {
    base: AdapterBase,
    api_key: String,
}

impl Sendgrid {
    /// PHP `__construct($apiKey)`.
    #[must_use]
    pub fn new(api_key: impl Into<String>) -> Self {
        Self {
            base: AdapterBase::default(),
            api_key: api_key.into(),
        }
    }

    fn process_email(&self, message: &Email) -> Result<ResponseData, MessagingError> {
        let mut personalizations: Vec<Value> = message
            .get_to()
            .iter()
            .map(|to| {
                let dest = if php_empty(to.name.as_deref()) {
                    json!({"email": to.email})
                } else {
                    json!({"email": to.email, "name": to.name})
                };
                json!({
                    "to": [dest],
                    "subject": message.get_subject(),
                })
            })
            .collect();

        if let Some(cc) = message.get_cc() {
            if !cc.is_empty() {
                for personalization in &mut personalizations {
                    let mut list = Vec::new();
                    for c in cc {
                        let mut entry = serde_json::Map::new();
                        entry.insert("email".into(), json!(c.email));
                        if !php_empty(c.name.as_deref()) {
                            entry.insert("name".into(), json!(c.name));
                        }
                        list.push(Value::Object(entry));
                    }
                    personalization["cc"] = Value::Array(list);
                }
            }
        }
        if let Some(bcc) = message.get_bcc() {
            if !bcc.is_empty() {
                for personalization in &mut personalizations {
                    let mut list = Vec::new();
                    for c in bcc {
                        let mut entry = serde_json::Map::new();
                        entry.insert("email".into(), json!(c.email));
                        if !php_empty(c.name.as_deref()) {
                            entry.insert("name".into(), json!(c.name));
                        }
                        list.push(Value::Object(entry));
                    }
                    personalization["bcc"] = Value::Array(list);
                }
            }
        }

        let mut attachments = Vec::new();
        if let Some(list) = message.get_attachments() {
            let mut size = 0u64;
            for attachment in list {
                size += std::fs::metadata(attachment.get_path())
                    .map(|m| m.len())
                    .unwrap_or(0);
            }
            if size > MAX_ATTACHMENT_BYTES {
                return Err(MessagingError::message(
                    "Attachments size exceeds the maximum allowed size of 25MB",
                ));
            }
            use base64::engine::general_purpose::STANDARD;
            use base64::Engine;
            for attachment in list {
                let data = std::fs::read(attachment.get_path()).unwrap_or_default();
                attachments.push(json!({
                    "content": STANDARD.encode(data),
                    "filename": attachment.get_name(),
                    "type": attachment.get_type(),
                    "disposition": "attachment",
                }));
            }
        }

        let mut body = json!({
            "personalizations": personalizations,
            "reply_to": {
                "name": message.get_reply_to_name(),
                "email": message.get_reply_to_email(),
            },
            "from": {
                "name": message.get_from_name(),
                "email": message.get_from_email(),
            },
            "content": [{
                "type": if message.is_html() { "text/html" } else { "text/plain" },
                "value": message.get_content(),
            }],
        });
        if !attachments.is_empty() {
            body["attachments"] = Value::Array(attachments);
        }

        let result = self.request_default(
            "POST",
            "https://api.sendgrid.com/v3/mail/send",
            &[
                format!("Authorization: Bearer {}", self.api_key),
                "Content-Type: application/json".into(),
            ],
            Some(body),
        );

        let mut response = Response::new(TYPE);
        if result.status_code == 202 {
            response.set_delivered_to(message.get_to().len() as i64);
            for to in message.get_to() {
                response.add_result(&to.email, "");
            }
        } else {
            let error = if let Some(s) = result.response.as_str() {
                s.to_string()
            } else if let Some(m) = result
                .response
                .get("errors")
                .and_then(|e| e.as_array())
                .and_then(|e| e.first())
                .and_then(|e| e.get("message"))
                .and_then(|v| v.as_str())
            {
                m.to_string()
            } else {
                "Unknown error".into()
            };
            for to in message.get_to() {
                response.add_result(&to.email, error.clone());
            }
        }
        Ok(response.to_array())
    }
}

impl Adapter for Sendgrid {
    fn get_name(&self) -> &'static str {
        "Sendgrid"
    }
    fn get_type(&self) -> &'static str {
        TYPE
    }
    fn get_message_type(&self) -> MessageKind {
        MessageKind::Email
    }
    fn get_max_messages_per_request(&self) -> usize {
        1000
    }
    fn base(&self) -> &AdapterBase {
        &self.base
    }
    fn process(&self, message: &dyn Message) -> Result<SendResult, MessagingError> {
        Ok(SendResult::Response(
            self.process_email(expect_email(message)?)?,
        ))
    }
}

File: packages/messaging/src/Adapter/Email/Sendgrid.php
<?php

namespace Utopia\Messaging\Adapter\Email;

use Utopia\Messaging\Adapter\Email as EmailAdapter;
use Utopia\Messaging\Messages\Email as EmailMessage;
use Utopia\Messaging\Response;

class Sendgrid extends EmailAdapter
{
    protected const NAME = 'Sendgrid';

    /**
     * @param  string  $apiKey Your Sendgrid API key to authenticate with the API.
     */
    public function __construct(private readonly string $apiKey)
    {
        parent::__construct();
    }

    /**
     * Get adapter name.
     */
    public function getName(): string
    {
        return static::NAME;
    }

    /**
     * Get max messages per request.
     */
    public function getMaxMessagesPerRequest(): int
    {
        return 1000;
    }

    /**
    * {@inheritdoc}
    *
    * Uses Sendgrid's personalization recipient variables to send multiple emails at once.
    *
    * @link https://www.twilio.com/docs/sendgrid/for-developers/sending-email/personalizations#-Sending-Two-Different-Emails-to-Two-Different-Groups-of-Recipients
    */
    protected function process(EmailMessage $message): array
    {
        $personalizations = array_map(
            fn (array $to): array => [
                'to' => [empty($to['name'])
                    ? ['email' => $to['email']]
                    : ['email' => $to['email'], 'name' => $to['name']]],
                'subject' => $message->getSubject(),
            ],
            $message->getTo(),
        );

        if (!\in_array($message->getCC(), [null, []], true)) {
            foreach ($personalizations as &$personalization) {
                foreach ($message->getCC() as $cc) {
                    $entry = ['email' => $cc['email']];
                    if (!empty($cc['name'])) {
                        $entry['name'] = $cc['name'];
                    }
                    $personalization['cc'][] = $entry;
                }
            }
            unset($personalization);
        }

        if (!\in_array($message->getBCC(), [null, []], true)) {
            foreach ($personalizations as &$personalization) {
                foreach ($message->getBCC() as $bcc) {
                    $entry = ['email' => $bcc['email']];
                    if (!empty($bcc['name'])) {
                        $entry['name'] = $bcc['name'];
                    }
                    $personalization['bcc'][] = $entry;
                }
            }
            unset($personalization);
        }

        $attachments = [];

        if (!\is_null($message->getAttachments())) {
            $size = 0;

            foreach ($message->getAttachments() as $attachment) {
                $size += filesize($attachment->getPath());
            }

            if ($size > self::MAX_ATTACHMENT_BYTES) {
                throw new \Exception('Attachments size exceeds the maximum allowed size of 25MB');
            }

            foreach ($message->getAttachments() as $attachment) {
                $attachments[] = [
                    'content' => base64_encode(file_get_contents($attachment->getPath())),
                    'filename' => $attachment->getName(),
                    'type' => $attachment->getType(),
                    'disposition' => 'attachment',
                ];
            }
        }

        $body = [
            'personalizations' => $personalizations,
            'reply_to' => [
                'name' => $message->getReplyToName(),
                'email' => $message->getReplyToEmail(),
            ],
            'from' => [
                'name' => $message->getFromName(),
                'email' => $message->getFromEmail(),
            ],
            'content' => [
                [
                    'type' => $message->isHtml() ? 'text/html' : 'text/plain',
                    'value' => $message->getContent(),
                ],
            ],
        ];

        if ($attachments !== []) {
            $body['attachments'] = $attachments;
        }

        if ($message->getHeaders() !== []) {
            $body['headers'] = $message->getHeaders();
        }

        $response = new Response($this->getType());
        $result = $this->request(
            method: 'POST',
            url: 'https://api.sendgrid.com/v3/mail/send',
            headers: [
                'Authorization: Bearer ' . $this->apiKey,
                'Content-Type: application/json',
            ],
            body: $body,
        );

        $statusCode = $result['statusCode'];

        if ($statusCode === 202) {
            $response->setDeliveredTo(\count($message->getTo()));
            foreach ($message->getTo() as $to) {
                $response->addResult($to['email']);
            }
        } else {
            foreach ($message->getTo() as $to) {
                if (\is_string($result['response'])) {
                    $response->addResult($to['email'], $result['response']);
                } elseif (!\is_null($result['response']['errors'][0]['message'] ?? null)) {
                    $response->addResult($to['email'], $result['response']['errors'][0]['message']);
                } else {
                    $response->addResult($to['email'], 'Unknown error');
                }
            }
        }

        return $response->toArray();
    }
}
